File: src/tryops/trace_envelope.py
from __future__ import annotations

import json
import re
from datetime import UTC, datetime
from typing import Any

SCHEMA_VERSION = "tryops.native_trace_log_envelope.v1"

_TRACE_ID = re.compile(r"^[0-9a-f]{32}$")
_SPAN_ID = re.compile(r"^[0-9a-f]{16}$")
_TRACE_FLAGS = re.compile(r"^[0-9a-f]{2}$")
_SENSITIVE_KEYS = {
    "prompt",
    "raw_prompt",
    "person_image_path",
    "garment_image_path",
    "output_image_path",
    "api_key",
    "authorization",
    "token",
    "secret",
}


def build_native_trace_log_envelope(
    event: dict[str, Any],
    *,
    structured_log: dict[str, Any] | None = None,
    language: str = "fastapi",
    runtime: str = "python",
    component: str = "api",
    service_name: str = "tryops-api",
    service_version: str = "0.1.0",
) -> dict[str, Any]:
    attributes = dict((structured_log or {}).get("attributes", {}))
    if not attributes:
        attributes = {
            "endpoint": event.get("endpoint"),
            "model_alias": event.get("model_alias"),
            "model_version": event.get("model_version"),
            "status": event.get("status"),
            "latency_ms": event.get("latency_ms"),
            "payload_metadata": event.get("payload_metadata", {}),
        }
    attributes = _drop_none(attributes)
    trace = event.get("trace", {}) if isinstance(event.get("trace"), dict) else {}
    trace_id = str(event.get("trace_id") or trace.get("trace_id") or "")
    span_id = str(event.get("span_id") or trace.get("span_id") or "")
    trace_flags = str(trace.get("trace_flags") or (structured_log or {}).get("trace_flags") or "01")
    traceparent = str(event.get("traceparent") or f"00-{trace_id}-{span_id}-{trace_flags}")

    return {
        "schema_version": SCHEMA_VERSION,
        "timestamp": str(event.get("timestamp") or datetime.now(UTC).isoformat()),
        "observed_timestamp": str((structured_log or {}).get("observed_timestamp") or datetime.now(UTC).isoformat()),
        "language": language,
        "runtime": runtime,
        "component": component,
        "event_name": str((structured_log or {}).get("event_name") or "tryops.api.request"),
        "severity_text": str((structured_log or {}).get("severity_text") or "INFO"),
        "severity_number": int((structured_log or {}).get("severity_number") or 9),
        "trace_id": trace_id,
        "span_id": span_id,
        "trace_flags": trace_flags,
        "traceparent": traceparent,
        "request_id": str(event.get("request_id") or attributes.get("request_id") or ""),
        "workload": str(event.get("workload") or attributes.get("workload") or ""),
        "resource": {
            "service.name": service_name,
            "service.version": service_version,
            "telemetry.sdk.language": language,
        },
        "attributes": attributes,
    }


def validate_native_trace_log_envelope(envelope: dict[str, Any]) -> list[str]:
    errors: list[str] = []
    required = [
        "schema_version",
        "timestamp",
        "observed_timestamp",
        "language",
        "runtime",
        "component",
        "event_name",
        "severity_text",
        "severity_number",
        "trace_id",
        "span_id",
        "trace_flags",
        "traceparent",
        "request_id",
        "workload",
        "resource",
        "attributes",
    ]
    for key in required:
        if key not in envelope:
            errors.append(f"missing {key}")
    if envelope.get("schema_version") != SCHEMA_VERSION:
        errors.append("schema_version mismatch")

    trace_id = str(envelope.get("trace_id", ""))
    span_id = str(envelope.get("span_id", ""))
    trace_flags = str(envelope.get("trace_flags", ""))
    traceparent = str(envelope.get("traceparent", ""))
    if not _TRACE_ID.fullmatch(trace_id) or _is_all_zero(trace_id):
        errors.append("invalid trace_id")
    if not _SPAN_ID.fullmatch(span_id) or _is_all_zero(span_id):
        errors.append("invalid span_id")
    if not _TRACE_FLAGS.fullmatch(trace_flags):
        errors.append("invalid trace_flags")
    if traceparent != f"00-{trace_id}-{span_id}-{trace_flags}":
        errors.append("traceparent does not match trace fields")

    resource = envelope.get("resource")
    if not isinstance(resource, dict):
        errors.append("resource must be an object")
    else:
        if not str(resource.get("service.name", "")).strip():
            errors.append("resource.service.name is required")
        if not str(resource.get("service.version", "")).strip():
            errors.append("resource.service.version is required")

    if not isinstance(envelope.get("attributes"), dict):
        errors.append("attributes must be an object")
    elif _has_sensitive_attribute(envelope["attributes"]):
        errors.append("attributes contain sensitive raw fields")
    if not str(envelope.get("event_name", "")).strip():
        errors.append("event_name is required")
    if int(envelope.get("severity_number") or 0) <= 0:
        errors.append("severity_number must be positive")
    return errors


def envelope_from_structured_log(record: dict[str, Any]) -> dict[str, Any]:
    envelope = record.get("native_envelope")
    if isinstance(envelope, dict):
        return envelope
    event = {
        "timestamp": record.get("timestamp"),
        "request_id": record.get("attributes", {}).get("request_id"),
        "workload": record.get("attributes", {}).get("workload"),
        "trace_id": record.get("trace_id"),
        "span_id": record.get("span_id"),
        "trace": {"trace_flags": record.get("trace_flags")},
        "traceparent": record.get("attributes", {}).get("traceparent"),
    }
    return build_native_trace_log_envelope(event, structured_log=record)


def _drop_none(value: dict[str, Any]) -> dict[str, Any]:
    return {key: item for key, item in value.items() if item is not None}


def _is_all_zero(value: str) -> bool:
    return all(ch == "0" for ch in value)


def _has_sensitive_attribute(value: Any) -> bool:
    if isinstance(value, dict):
        for key, item in value.items():
            if str(key).lower() in _SENSITIVE_KEYS:
                return True
            if _has_sensitive_attribute(item):
                return True
        return False
    if isinstance(value, list):
        return any(_has_sensitive_attribute(item) for item in value)
    if isinstance(value, str):
        serialized = value.lower()
        if "bearer " in serialized or "api_key" in serialized or "secret prompt" in serialized:
            return True
    return False


def envelope_json(envelope: dict[str, Any]) -> str:
    return json.dumps(envelope, sort_keys=True)

File: native/rust/tryops-gateway/src/trace_envelope.rs
use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

use crate::trace_context::TraceContext;

pub(crate) const NATIVE_TRACE_LOG_SCHEMA: &str = "tryops.native_trace_log_envelope.v1";

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub(crate) struct NativeTraceLogEnvelope {
    pub(crate) schema_version: String,
    pub(crate) timestamp: String,
    pub(crate) observed_timestamp: String,
    pub(crate) language: String,
    pub(crate) runtime: String,
    pub(crate) component: String,
    pub(crate) event_name: String,
    pub(crate) severity_text: String,
    pub(crate) severity_number: u8,
    pub(crate) trace_id: String,
    pub(crate) span_id: String,
    pub(crate) trace_flags: String,
    pub(crate) traceparent: String,
    pub(crate) request_id: String,
    pub(crate) workload: String,
    pub(crate) resource: BTreeMap<String, String>,
    pub(crate) attributes: BTreeMap<String, String>,
}

impl NativeTraceLogEnvelope {
    pub(crate) fn gateway_proxy_request(
        trace_context: &TraceContext,
        request_id: &str,
        endpoint: &str,
        method: &str,
        service_version: &str,
    ) -> Self {
        let timestamp = unix_timestamp_millis();
        let mut resource = BTreeMap::new();
        resource.insert("service.name".to_string(), "tryops-gateway".to_string());
        resource.insert("service.version".to_string(), service_version.to_string());
        resource.insert("telemetry.sdk.language".to_string(), "rust".to_string());

        let mut attributes = BTreeMap::new();
        attributes.insert("endpoint".to_string(), endpoint.to_string());
        attributes.insert("method".to_string(), method.to_string());
        attributes.insert("status".to_string(), "forwarded".to_string());

        Self {
            schema_version: NATIVE_TRACE_LOG_SCHEMA.to_string(),
            timestamp: timestamp.clone(),
            observed_timestamp: timestamp,
            language: "rust".to_string(),
            runtime: "tokio-axum".to_string(),
            component: "edge-gateway".to_string(),
            event_name: "tryops.gateway.proxy.request".to_string(),
            severity_text: "INFO".to_string(),
            severity_number: 9,
            trace_id: trace_context.trace_id().to_string(),
            span_id: trace_context.span_id().to_string(),
            trace_flags: trace_context.trace_flags().to_string(),
            traceparent: trace_context.traceparent().to_string(),
            request_id: request_id.to_string(),
            workload: workload_from_endpoint(endpoint).to_string(),
            resource,
            attributes,
        }
    }

    #[cfg(test)]
    pub(crate) fn validate(&self) -> Vec<String> {
        let mut errors = Vec::new();
        if self.schema_version != NATIVE_TRACE_LOG_SCHEMA {
            errors.push("schema_version mismatch".to_string());
        }
        if !is_hex_len(&self.trace_id, 32) || is_all_zero(&self.trace_id) {
            errors.push("invalid trace_id".to_string());
        }
        if !is_hex_len(&self.span_id, 16) || is_all_zero(&self.span_id) {
            errors.push("invalid span_id".to_string());
        }
        if !is_hex_len(&self.trace_flags, 2) {
            errors.push("invalid trace_flags".to_string());
        }
        if self.traceparent != format!("00-{}-{}-{}", self.trace_id, self.span_id, self.trace_flags)
        {
            errors.push("traceparent does not match trace fields".to_string());
        }
        if self
            .resource
            .get("service.name")
            .map_or(true, |value| value.is_empty())
        {
            errors.push("resource.service.name is required".to_string());
        }
        if self
            .resource
            .get("service.version")
            .map_or(true, |value| value.is_empty())
        {
            errors.push("resource.service.version is required".to_string());
        }
        if self.event_name.is_empty() {
            errors.push("event_name is required".to_string());
        }
        if self.severity_number == 0 {
            errors.push("severity_number must be positive".to_string());
        }
        errors
    }

    pub(crate) fn with_outcome(mut self, status_code: u16, latency_ms: f64) -> Self {
        self.attributes
            .insert("status".to_string(), status_code.to_string());
        self.attributes
            .insert("latency_ms".to_string(), format!("{latency_ms:.3}"));
        if status_code >= 500 {
            self.severity_text = "ERROR".to_string();
            self.severity_number = 17;
        } else if status_code >= 400 {
            self.severity_text = "WARN".to_string();
            self.severity_number = 13;
        }
        self
    }
}

fn workload_from_endpoint(endpoint: &str) -> &'static str {
    if endpoint.contains("/llm/") {
        "llm"
    } else if endpoint.contains("/vton/") {
        "vton"
    } else if endpoint.contains("/quota/") {
        "quota"
    } else {
        "platform"
    }
}

fn unix_timestamp_millis() -> String {
    let millis = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis();
    format!("{millis}")
}

#[cfg(test)]
fn is_hex_len(value: &str, length: usize) -> bool {
    value.len() == length
        && value
            .bytes()
            .all(|byte| byte.is_ascii_digit() || (b'a'..=b'f').contains(&byte))
}

#[cfg(test)]
fn is_all_zero(value: &str) -> bool {
    value.bytes().all(|byte| byte == b'0')
}

#[cfg(test)]
mod tests {
    use axum::http::{HeaderMap, HeaderValue};

    use crate::trace_context::trace_context_from_headers;

    use super::*;

    #[test]
    fn gateway_envelope_matches_shared_trace_contract() {
        let mut headers = HeaderMap::new();
        headers.insert(
            "traceparent",
            HeaderValue::from_static("00-11111111111111111111111111111111-2222222222222222-01"),
        );
        let context = trace_context_from_headers(&headers);

        let envelope = NativeTraceLogEnvelope::gateway_proxy_request(
            &context,
            "req-rust",
            "/v1/llm/generate",
            "POST",
            "0.1.0",
        );

        assert_eq!(envelope.schema_version, NATIVE_TRACE_LOG_SCHEMA);
        assert_eq!(envelope.trace_id, "11111111111111111111111111111111");
        assert_eq!(envelope.workload, "llm");
        assert!(envelope.traceparent.ends_with("-01"));
        assert_eq!(envelope.validate(), Vec::<String>::new());
    }

    #[test]
    fn gateway_envelope_records_outcome_without_breaking_contract() {
        let mut headers = HeaderMap::new();
        headers.insert(
            "traceparent",
            HeaderValue::from_static("00-11111111111111111111111111111111-2222222222222222-01"),
        );
        let context = trace_context_from_headers(&headers);
        let envelope = NativeTraceLogEnvelope::gateway_proxy_request(
            &context,
            "req-rust",
            "/v1/llm/generate",
            "POST",
            "0.1.0",
        )
        .with_outcome(429, 12.3456);

        assert_eq!(envelope.attributes["status"], "429");
        assert_eq!(envelope.attributes["latency_ms"], "12.346");
        assert_eq!(envelope.severity_text, "WARN");
        assert_eq!(envelope.validate(), Vec::<String>::new());
    }

    #[test]
    fn validation_rejects_invalid_trace_ids() {
        let mut headers = HeaderMap::new();
        headers.insert(
            "traceparent",
            HeaderValue::from_static("00-11111111111111111111111111111111-2222222222222222-01"),
        );
        let context = trace_context_from_headers(&headers);
        let mut envelope = NativeTraceLogEnvelope::gateway_proxy_request(
            &context,
            "req-rust",
            "/v1/vton/infer",
            "POST",
            "0.1.0",
        );
        envelope.trace_id = "00000000000000000000000000000000".to_string();

        assert!(envelope
            .validate()
            .contains(&"invalid trace_id".to_string()));
    }
}
